### DISCUSSION TITLE
Prisma and Next.js Static Site Generation

### DISCUSSION THREAD

**Discussion by adlradlr** [prisma/prisma#5537] [Q&A]
### Prisma and Next.js Static Site Generation
I've been debugging some issues for this particular setup and thought I'd share my findings for anyone else who may run into similar issues.

_**Context**: I'm using Prisma with a Postgres DB in a Next.js application that uses SSG/ISR. App size ~130 pages_ 

So far I have only been using Prisma inside of getStaticPaths and getStaticProps with incremental static regeneration enabled (no Next.js serverless API routes) and have been running into a lot of issues with connection limits. Once deployed/built the number of active connections would quickly hit the limit of 20 and also remain at 20 for around 1 hour before dropping off back down to 0.
<img width="1182" alt="Screenshot 2021-02-06 at 21 44 59" src="https://user-images.githubusercontent.com/26056727/107212265-9bb6f300-69fe-11eb-9aab-10583a10e4f0.png">
<img width="1171" alt="Screenshot 2021-02-06 at 21 45 13" src="https://user-images.githubusercontent.com/26056727/107212275-9fe31080-69fe-11eb-9403-3d3bc7640df4.png">

This was causing a lot of issues because this effectively locked me out of the app. It then became impossible to redeploy the app with fixes or changes for the next hour without stopping the production build, as every new build attempt would fail due to too many active connections already.

`Unhandled error during request: PrismaClientInitializationError2 [PrismaClientInitializationError]: Error querying the database: db error: FATAL: remaining connection slots are reserved for non-replication superuser connections`

**What I've tried**:

**Enabling/Disabling Incremental Static Regeneration**:
- This doesn't seem to have any negative effect.

**Initializing PrismaClient once**:
- This was something I initially missed and has helped. Only create one instance of PrismaClient and import it into the pages where getStaticPaths/getStaticProps are being used.
```javascript
// prisma.js
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()
export default prisma


// [post].js
import prisma from '../prisma.js'

...

export const getStaticProps = async () => {
  const post = await prisma.posts.findUnique()
}
```

**Explicitly disconnecting Prisma after every use**:
- This has helped in ensuring the active connections don't skyrocket to the max limit almost instantly.

```javascript
// [post].js
import prisma from '../prisma.js'

...

export const getStaticProps = async () => {
  const post = await prisma.posts.findUnique()

  ...

  await prisma.$disconnect()
}
```

**Setting DB connection limit**:
-  Adding a connection limit parameter to the DB connection URL.
`DBURL?connection_limit=1`

I'm currently employing all of these tips, but setting the connection limit to 1 seems to be the 'final piece to the puzzle'.  Active connections still go far above the new set limit of 1 initially after deploying to production. However, I think this is now down to intended behaviours from ISR/Vercel.

<img width="1095" alt="Screenshot 2021-02-08 at 12 00 36" src="https://user-images.githubusercontent.com/26056727/107216848-429e8d80-6a05-11eb-8cf2-6607502aaa73.png">
<img width="1094" alt="Screenshot 2021-02-08 at 12 00 51" src="https://user-images.githubusercontent.com/26056727/107216880-4d592280-6a05-11eb-89e8-39609799f804.png">

As stated, it's now working as intended. So while I aim to just leave this as feedback for anyone who encounters similar issues in the future, feel free to let me know any issues/improvements. -Thanks

**Comment by laravelman** [comment]
Damn it helped! Saved my day! Thanks.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by Bogd1y** [comment]
absolute legend, thanks
